File: java/PalindromeLinkedList234.java
/*
Given the head of a singly linked list, return true if it is a palindrome or false otherwise.
*/
public class PalindromeLinkedList234 {
    public static void main(String[] args) {
//        1 2 3 3 2 1
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(3, new ListNode(2, new ListNode(1))))));
        System.out.println(isPalindrome(head));

        head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(2, new ListNode(1)))));
        System.out.println(isPalindrome(head));

        head = new ListNode(1, new ListNode(2, new ListNode(2, new ListNode(1))));
        System.out.println(isPalindrome(head));

        head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        System.out.println(isPalindrome(head));
    }

    public static boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) return true;
        ListNode slow = head, fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode end = null, next;
        while (slow != null) {
            next = slow.next;
            slow.next = end;
            end = slow;
            slow = next;
        }
        while (head != null && end != null) {
            if (head.val != end.val) return false;
            head = head.next;
            end = end.next;
        }
        return true;
    }

    public static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}

File: python/PalindromeLinkedList234.py
# Given the head of a singly linked list, return true if it is a palindrome or false otherwise.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def isPalindrome(head: Optional[ListNode]) -> bool:
    if not head or not head.next:
        return True
    slow, fast = head, head
    while fast.next and fast.next.next:
        slow, fast = slow.next, fast.next.next
    end = None
    while slow:
        slow.next, end, slow = end, slow, slow.next
    while head and end:
        if head.val != end.val:
            return False
        head, end = head.next, end.next
    return True


if __name__ == "__main__":
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(2)
    head_outer.next.next.next.next = ListNode(1)
    print(isPalindrome(head_outer))
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(4)
    head_outer.next.next.next.next = ListNode(5)
    print(isPalindrome(head_outer))
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(3)
    head_outer.next.next.next.next = ListNode(2)
    head_outer.next.next.next.next.next = ListNode(1)
    print(isPalindrome(head_outer))
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(4)
    head_outer.next.next.next.next = ListNode(4)
    head_outer.next.next.next.next.next = ListNode(3)
    head_outer.next.next.next.next.next.next = ListNode(2)
    head_outer.next.next.next.next.next.next.next = ListNode(1)
    print(isPalindrome(head_outer))
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(4)
    head_outer.next.next.next.next = ListNode(5)
    head_outer.next.next.next.next.next = ListNode(4)
    head_outer.next.next.next.next.next.next = ListNode(3)
    head_outer.next.next.next.next.next.next.next = ListNode(2)
    head_outer.next.next.next.next.next.next.next.next = ListNode(1)
    print(isPalindrome(head_outer))
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(4)
    head_outer.next.next.next.next = ListNode(5)
    head_outer.next.next.next.next.next = ListNode(5)
    head_outer.next.next.next.next.next.next = ListNode(4)
    head_outer.next.next.next.next.next.next.next = ListNode(3)
    head_outer.next.next.next.next.next.next.next.next = ListNode(2)
    head_outer.next.next.next.next.next.next.next.next.next = ListNode(1)
    print(isPalindrome(head_outer))
